src/wasm.rs

```diff
@@ -1,5 +1,51 @@
 use alloc::string::String;
 
+use js_sys::Object;
+use wasm_bindgen::{prelude::*, JsCast, JsValue};
+
+#[derive(Clone)]
+enum GlobalType {
+    Window(web_sys::Window),
+    Worker(web_sys::WorkerGlobalScope),
+}
+
+/// Returns a handle to the global scope object.
+///
+/// Simplified version of https://github.com/rustwasm/wasm-bindgen/blob/main/crates/js-sys/src/lib.rs,
+/// which we can't use directly because it discards information about how it
+/// retrieved the global.
+fn global() -> Option<GlobalType> {
+    #[wasm_bindgen]
+    extern "C" {
+        type Global;
+
+        #[wasm_bindgen(getter, catch, static_method_of = Global, js_class = window, js_name = window)]
+        fn get_window() -> Result<Object, JsValue>;
+
+        #[wasm_bindgen(getter, catch, static_method_of = Global, js_class = self, js_name = self)]
+        fn get_self() -> Result<Object, JsValue>;
+    }
+
+    if let Ok(window) = Global::get_window() {
+        Some(GlobalType::Window(
+            window
+                .dyn_into::<web_sys::Window>()
+                .expect("window was not a Window"),
+        ))
+    } else if let Ok(worker) = Global::get_self() {
+        Some(GlobalType::Worker(
+            worker
+                .dyn_into::<web_sys::WorkerGlobalScope>()
+                .expect("self was not a WorkerGlobalScope"),
```

Comment:
You're totally right that we're making some assumptions about things. `self` is actually defined in `window`, but that wouldn't allow us to distinguish between a page and worker, which is why we check `window` first. The reason we need to distinguish between them is that a page has a `Navigator` and a worker has a `WorkerNavigator`, and these are similar but not the same.

I assume your proposal (with a few tweaks to make it real) is something like this:

```rust
pub(crate) fn get() -> Option<String> {
    #[wasm_bindgen]
    extern "C" {
        type Global;

        #[wasm_bindgen(getter, catch, static_method_of = Global, js_class = globalThis, js_name = navigator)]
        fn get_navigator() -> Result<Option<web_sys::Navigator>, JsValue>;
    }

    Global::get_navigator().ok()??.language()
}
```

That does work, but:

- I've had to use `globalThis`, since `global` isn't defined in a page. This unfortunately wouldn't work for us because we currently support browsers that don't implement this.
- We're always casting to `Navigator`, which isn't the real type. It seems like wasm-bindgen lets us get away with this (and it works because both have a language function) but doesn't feel great. If someone copied our code to use for other things on navigator it might not work for them.

We could use `self` instead of `globalThis`, which is defined everywhere, and decide we're comfortable with abusing the Navigator type. I _think_ this would work, although there are definitely caveats to using `self` which is the whole reason `globalThis` was proposed. I don't know for definite if that would be a problem for us (I can't think of any) but it does make me a little reluctant to depend on self given that I know that's not a thing people usually rely on in windows.

So in summary: I definitely share your goal of simplifying things a little, but it's tough.

Side note: For your snippet you moved everything in to `get` which definitely simplifies things. Would we rather do this? I'm happy to do that but it does inline a lot more and make it less reusable in the future.

Reply:
* Moving. I only did that for the sake of a simple snippet.
* Excellent point about `globalThis`. I didn't realize we couldn't use that everywhere.
* If think it's safe to cast to the wrong `Navigator` object because we only call the one method; for additional safety, we could forgo the `Navigator` object completely and define our own class with just a single method on it

Overall, I do like the strong typing in your change. My primary suggestion would be tweaking it a bit so that it doesn't panic in non-browser environments.

Reply:
Alright - having discussed this offline a bit, we agreed to keep this approach and actually panic more consistently. The idea is that it's better to throw when you are running the library in an unsupported environment then silently fail and pretend like things are working.